API for more ergonomic construction of `RecordBatchOptions`

**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
When creating a `RecordBatchOption`s object it would be nice to use a standard Rust patterns of either

A builder interface :

```rust
let options = RecordBatchOptions::new()
  .with_rows_count(Some(row_count))
```

Or `Default::default()` for `pub` structs:
```rust
let options = RecordBatchOptions{
      row_count:Some(row_count),
      ..Default::default()
    };
```

However, there is no builder interface provided and `RecordBatchOptions` is marked as non-exhaustive so you get this error 

```
error[E0639]: cannot create non-exhaustive struct using struct expression
   --> /home/isidentical/projects/arrow-datafusion/datafusion/core/src/physical_plan/coalesce_batches.rs:295:19
    |
295 |       let options = RecordBatchOptions {
    |  ___________________^
296 | |         row_count: Some(row_count),
297 | |         ..Default::default()
298 | |     };
    | |_____^

For more information about this error, try `rustc --explain E0639`.
```

This means you have to use `mut` like:

```rust
  let mut options = RecordBatchOptions::default();
  options.row_count = Some(row_count);
```

**Describe the solution you'd like**
Add a builder interface for RecordBatchOptions


**Additional context**

See comments from @isidentical https://github.com/apache/arrow-datafusion/pull/3439#discussion_r968556985

--- a/arrow/src/ipc/reader.rs
+++ b/arrow/src/ipc/reader.rs
@@ -31,7 +31,7 @@ use crate::compute::cast;
 use crate::datatypes::{DataType, Field, IntervalUnit, Schema, SchemaRef, UnionMode};
 use crate::error::{ArrowError, Result};
 use crate::ipc;
-use crate::record_batch::{RecordBatch, RecordBatchOptionsBuilder, RecordBatchReader};
+use crate::record_batch::{RecordBatch, RecordBatchOptions, RecordBatchReader};
 
 use crate::ipc::compression::CompressionCodec;
 use ipc::CONTINUATION_MARKER;
@@ -578,9 +578,7 @@ pub fn read_record_batch(
     let mut node_index = 0;
     let mut arrays = vec![];
 
-    let options = RecordBatchOptionsBuilder::new()
-        .row_count(batch.length() as usize)
-        .build();
+    let options = RecordBatchOptions::new().with_row_count(Some(batch.length() as usize));
 
     if let Some(projection) = projection {
         // project fields
@@ -1691,10 +1689,9 @@ mod tests {
     #[test]
     fn test_no_columns_batch() {
         let schema = Arc::new(Schema::new(vec![]));
-        let options = RecordBatchOptionsBuilder::new()
-            .match_field_names(true)
-            .row_count(10)
-            .build();
+        let options = RecordBatchOptions::new()
+            .with_match_field_names(true)
+            .with_row_count(Some(10));
         let input_batch =
             RecordBatch::try_new_with_options(schema, vec![], &options).unwrap();
         let output_batch = roundtrip_ipc_stream(&input_batch);

--- a/arrow/src/ipc/reader.rs
+++ b/arrow/src/ipc/reader.rs
@@ -31,7 +31,7 @@ use crate::compute::cast;
 use crate::datatypes::{DataType, Field, IntervalUnit, Schema, SchemaRef, UnionMode};
 use crate::error::{ArrowError, Result};
 use crate::ipc;
-use crate::record_batch::{RecordBatch, RecordBatchOptions, RecordBatchReader};
+use crate::record_batch::{RecordBatch, RecordBatchOptionsBuilder, RecordBatchReader};
 
 use crate::ipc::compression::CompressionCodec;
 use ipc::CONTINUATION_MARKER;
@@ -578,10 +578,9 @@ pub fn read_record_batch(
     let mut node_index = 0;
     let mut arrays = vec![];
 
-    let options = RecordBatchOptions {
-        row_count: Some(batch.length() as usize),
-        ..Default::default()
-    };
+    let options = RecordBatchOptionsBuilder::new()
+        .row_count(batch.length() as usize)
+        .build();
 
     if let Some(projection) = projection {
         // project fields
@@ -1692,10 +1691,10 @@ mod tests {
     #[test]
     fn test_no_columns_batch() {
         let schema = Arc::new(Schema::new(vec![]));
-        let options = RecordBatchOptions {
-            match_field_names: true,
-            row_count: Some(10),
-        };
+        let options = RecordBatchOptionsBuilder::new()
+            .match_field_names(true)
+            .row_count(10)
+            .build();
         let input_batch =
             RecordBatch::try_new_with_options(schema, vec![], &options).unwrap();
         let output_batch = roundtrip_ipc_stream(&input_batch);

--- a/arrow/src/record_batch.rs
+++ b/arrow/src/record_batch.rs
@@ -963,7 +963,7 @@ mod tests {
             .match_field_names(false)
             .row_count(20)
             .build();
-        assert_eq!(options.match_field_names, false);
+        assert!(!options.match_field_names);
         assert_eq!(options.row_count.unwrap(), 20)
     }
 }

--- a/arrow/src/record_batch.rs
+++ b/arrow/src/record_batch.rs
@@ -80,7 +80,7 @@ impl RecordBatch {
     /// # }
     /// ```
     pub fn try_new(schema: SchemaRef, columns: Vec<ArrayRef>) -> Result<Self> {
-        let options = RecordBatchOptionsBuilder::new().build();
+        let options = RecordBatchOptions::new();
         Self::try_new_impl(schema, columns, &options)
     }
 
@@ -412,47 +412,31 @@ pub struct RecordBatchOptions {
     /// Optional row count, useful for specifying a row count for a RecordBatch with no columns
     pub row_count: Option<usize>,
 }
-/// Builder to create new RecordBatchOptions object
-#[derive(Debug)]
-pub struct RecordBatchOptionsBuilder {
-    /// Match field names of structs and lists. If set to `true`, the names must match.
-    match_field_names: bool,
 
-    /// Optional row count, useful for specifying a row count for a RecordBatch with no columns
-    row_count: Option<usize>,
-}
-impl Default for RecordBatchOptionsBuilder {
-    fn default() -> Self {
-        Self::new()
-    }
-}
-
-impl RecordBatchOptionsBuilder {
+impl RecordBatchOptions {
     pub fn new() -> Self {
         Self {
             match_field_names: true,
             row_count: None,
         }
     }
-    pub fn match_field_names(mut self, match_field_names: bool) -> Self {
+    //Sets the row_count of RecordBatchOptions and returns self
+    pub fn with_row_count(mut self, row_count: Option<usize>) -> Self {
+        self.row_count = row_count;
+        self
+    }
+    //Sets the match_field_names of RecordBatchOptions and returns self
+    pub fn with_match_field_names(mut self, match_field_names: bool) -> Self {
         self.match_field_names = match_field_names;
         self
     }
-    pub fn row_count(mut self, row_count: usize) -> Self {
-        self.row_count = Some(row_count);
-        self
-    }
-    pub fn build(self) -> RecordBatchOptions {
-        let Self {
-            match_field_names,
-            row_count,
-        } = self;
-        RecordBatchOptions {
-            match_field_names,
-            row_count,
-        }
+}
+impl Default for RecordBatchOptions {
+    fn default() -> Self {
+        Self::new()
     }
 }
+/// Builder to create new RecordBatchOptions object
 impl From<&StructArray> for RecordBatch {
     /// Create a record batch from struct array, where each field of
     /// the `StructArray` becomes a `Field` in the schema.
@@ -932,7 +916,7 @@ mod tests {
             .to_string()
             .contains("must either specify a row count or at least one column"));
 
-        let options = RecordBatchOptionsBuilder::new().row_count(10).build();
+        let options = RecordBatchOptions::new().with_row_count(Some(10));
 
         let ok =
             RecordBatch::try_new_with_options(schema.clone(), vec![], &options).unwrap();
@@ -958,11 +942,10 @@ mod tests {
         assert_eq!("Invalid argument error: Column 'a' is declared as non-nullable but contains null values", format!("{}", maybe_batch.err().unwrap()));
     }
     #[test]
-    fn test_record_batch_options_builder() {
-        let options = RecordBatchOptionsBuilder::new()
-            .match_field_names(false)
-            .row_count(20)
-            .build();
+    fn test_record_batch_options() {
+        let options = RecordBatchOptions::new()
+            .with_match_field_names(false)
+            .with_row_count(Some(20));
         assert!(!options.match_field_names);
         assert_eq!(options.row_count.unwrap(), 20)
     }

--- a/arrow/src/record_batch.rs
+++ b/arrow/src/record_batch.rs
@@ -425,7 +425,7 @@ impl RecordBatchOptions {
         self.row_count = row_count;
         self
     }
-    //Sets the match_field_names of RecordBatchOptions and returns self
+    /// Sets the match_field_names of RecordBatchOptions and returns self
     pub fn with_match_field_names(mut self, match_field_names: bool) -> Self {
         self.match_field_names = match_field_names;
         self

--- a/arrow/src/record_batch.rs
+++ b/arrow/src/record_batch.rs
@@ -436,7 +436,6 @@ impl Default for RecordBatchOptions {
         Self::new()
     }
 }
-/// Builder to create new RecordBatchOptions object
 impl From<&StructArray> for RecordBatch {
     /// Create a record batch from struct array, where each field of
     /// the `StructArray` becomes a `Field` in the schema.

--- a/arrow/src/record_batch.rs
+++ b/arrow/src/record_batch.rs
@@ -421,6 +421,11 @@ pub struct RecordBatchOptionsBuilder {
     /// Optional row count, useful for specifying a row count for a RecordBatch with no columns
     row_count: Option<usize>,
 }
+impl Default for RecordBatchOptionsBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
 
 impl RecordBatchOptionsBuilder {
     pub fn new() -> Self {

--- a/arrow/src/record_batch.rs
+++ b/arrow/src/record_batch.rs
@@ -420,7 +420,7 @@ impl RecordBatchOptions {
             row_count: None,
         }
     }
-    //Sets the row_count of RecordBatchOptions and returns self
+    /// Sets the row_count of RecordBatchOptions and returns self
     pub fn with_row_count(mut self, row_count: Option<usize>) -> Self {
         self.row_count = row_count;
         self

--- a/arrow/src/record_batch.rs
+++ b/arrow/src/record_batch.rs
@@ -80,7 +80,7 @@ impl RecordBatch {
     /// # }
     /// ```
     pub fn try_new(schema: SchemaRef, columns: Vec<ArrayRef>) -> Result<Self> {
-        let options = RecordBatchOptions::default();
+        let options = RecordBatchOptionsBuilder::new().build();
         Self::try_new_impl(schema, columns, &options)
     }
 
@@ -412,16 +412,42 @@ pub struct RecordBatchOptions {
     /// Optional row count, useful for specifying a row count for a RecordBatch with no columns
     pub row_count: Option<usize>,
 }
+/// Builder to create new RecordBatchOptions object
+#[derive(Debug)]
+pub struct RecordBatchOptionsBuilder {
+    /// Match field names of structs and lists. If set to `true`, the names must match.
+    match_field_names: bool,
 
-impl Default for RecordBatchOptions {
-    fn default() -> Self {
+    /// Optional row count, useful for specifying a row count for a RecordBatch with no columns
+    row_count: Option<usize>,
+}
+
+impl RecordBatchOptionsBuilder {
+    pub fn new() -> Self {
         Self {
             match_field_names: true,
             row_count: None,
         }
     }
+    pub fn match_field_names(mut self, match_field_names: bool) -> Self {
+        self.match_field_names = match_field_names;
+        self
+    }
+    pub fn row_count(mut self, row_count: usize) -> Self {
+        self.row_count = Some(row_count);
+        self
+    }
+    pub fn build(self) -> RecordBatchOptions {
+        let Self {
+            match_field_names,
+            row_count,
+        } = self;
+        RecordBatchOptions {
+            match_field_names,
+            row_count,
+        }
+    }
 }
-
 impl From<&StructArray> for RecordBatch {
     /// Create a record batch from struct array, where each field of
     /// the `StructArray` becomes a `Field` in the schema.
@@ -901,10 +927,7 @@ mod tests {
             .to_string()
             .contains("must either specify a row count or at least one column"));
 
-        let options = RecordBatchOptions {
-            row_count: Some(10),
-            ..Default::default()
-        };
+        let options = RecordBatchOptionsBuilder::new().row_count(10).build();
 
         let ok =
             RecordBatch::try_new_with_options(schema.clone(), vec![], &options).unwrap();
@@ -928,5 +951,14 @@ mod tests {
             vec![Arc::new(Int32Array::from(vec![Some(1), None]))],
         );
         assert_eq!("Invalid argument error: Column 'a' is declared as non-nullable but contains null values", format!("{}", maybe_batch.err().unwrap()));
+    }
+    #[test]
+    fn test_record_batch_options_builder() {
+        let options = RecordBatchOptionsBuilder::new()
+            .match_field_names(false)
+            .row_count(20)
+            .build();
+        assert_eq!(options.match_field_names, false);
+        assert_eq!(options.row_count.unwrap(), 20)
     }
 }

